R/rdeephaven/R/helper_functions.R

```diff
@@ -1,18 +1,86 @@
-.verify_string <- function(arg_name, string_candidate) {
-  if (class(string_candidate)[[1]] != "character") {
-    stop(paste0("'", arg_name, "' must be passed as a single string. Got an object of class ", class(string_candidate)[[1]], " instead."))
-  } else if (length(string_candidate) != 1) {
-    stop(paste0("'", arg_name, "' must be passed as a single string. Got a character vector of length ", length(string_candidate), " instead."))
+first_class <- function(arg) {
+  return(class(arg)[[1]])
+}
+
+verify_type <- function(arg_name, candidate, required_type, message_type_name, is_scalar) {
+  if (first_class(candidate) == "list") {
+    if (any(lapply(candidate, first_class) != required_type)) {
+      stop(paste0("'", arg_name, "' must be a ", message_type_name, ", or a vector of ", message_type_name, "s. Got a vector with at least one element that is not a ", message_type_name, " instead."))
+    }
+  } else {
+    if (first_class(candidate) != required_type) {
+      if (is_scalar) {
+        stop(paste0("'", arg_name, "' must be passed as a single ", message_type_name, ". Got an object of class ", first_class(candidate), " instead."))
+      } else {
+        stop(paste0("'", arg_name, "' must be passed as a ", message_type_name, " or a vector of ", message_type_name, "s. Got an object of class ", first_class(candidate), " instead."))
+      }
+    }
+  }
+
+  if (is_scalar) {
+    if (length(candidate) != 1) {
+      stop(paste0("'", arg_name, "' must be passed as a single ", message_type_name, ". Got a ", message_type_name, " vector of length ", length(candidate), " instead."))
+    }
+  }
+}
+
+# does not attempt to verify that candidate is numeric
+verify_in_range <- function(arg_name, candidate, message, a = NULL, b = NULL, a_open = TRUE, b_open = TRUE) {
```

Comment:
I had that in here in a previous commit, and got rid of it because I thought the error messages were cleaner with `nonnegative` and `positive` instead of `[0, inf)` and `(0, inf)`, since this function is currently only used in `verify_nonnegative_int` and `verify_positive_int`.

Reply:
You could create those messages automatically.

Reply:
I feel we have a lot more flexibility just passing the messages directly. If you I create stuff internally I can do that.